src/kernel.c

```diff
@@ -0,0 +1,642 @@
+	for (int i=0;i<STACK_GUARD/8;i++)
+		s[i] = STACK_COOKIE;
+
+	// The main thread is the currently scheduled one.
+	th_cur = &th_main;
+
+	// Allocate the idle thread
+	th_idle = kthread_new("idle", 4096, -1, __kthread_idle, 0);
+
+	__kernel = true;
+	return th_cur;
+}
+
+void kernel_close(void)
+{
+	assert(__kernel);
+	assertf(th_cur == &th_main, "kthread_close can only be called from main thread");
+
+	// Kill the idle thread to release the memory
+	kthread_kill(th_idle);
+	th_idle = NULL;
+
+	assertf(th_count == 1, "not all threads were killed");
+
+	th_cur = NULL;
+	__kernel = false;
+	__isr_force_schedule = false;
+}
+
+kthread_t* kthread_new(const char *name, int stack_size, int8_t pri, void (*user_entry)(void*), void *user_data)
+{
+	assert((stack_size % 8) == 0);
+
+	// Allocate memory for the stack, the stack guard area, and the kthread_t
+	// structure
+	//
+	// |             |                        |             |
+	// | Stack guard |   Stack                | kthread_t   |
+	// |             |                        |             |
+	//
+	//
+	// For the kthread structure, we will use the final part of
+	// the buffer: since the stack grows downward, this makes sure that
+	// overflowing the stack doesn't corrupt the thread structure, which might
+	// make debugging more difficult and might even cause the kernel to crash.
+	void *thmem = malloc(STACK_GUARD + stack_size + sizeof(kthread_t));
+	assertf(thmem, "out of free memory");
+	kthread_t *th = thmem + STACK_GUARD + stack_size;
+
+	// Initialize the thread structure
+	memset(th, 0, sizeof(kthread_t));
+	th->stack = thmem;
+
+	th->name = name;
+	th->user_entry = user_entry;
+	th->user_data = user_data;
+	th->pri = pri;
+
+	// Initialize the stack guard
+	uint64_t *s = (uint64_t*)th->stack;
+	for (int i=0;i<STACK_GUARD/8;i++)
+		s[i] = STACK_COOKIE;
+
+	// Top of the stack is the end of the stack area, so where the kthread_t
+	// structure is. Remember that the stack grows downward.
+	void *top_stack = (void*)th;
+
+	// Put the initial regblock at the top of the stack. This is the
+	// block that will be popped when the thread is first scheduled, and needs to
+	// initialize all the registers and jump to the entry point.
+	th->stack_state = (reg_block_t*)(top_stack - sizeof(reg_block_t));
+	assert(((uint32_t)th->stack_state & 7) == 0);
+	memset(th->stack_state, 0, sizeof(reg_block_t));
+
+	// EPC: __thread_boot function (which will then invoke the user entry point)
+	th->stack_state->epc = (uint32_t)&__kthread_boot;
+
+	// SR: start in exception mode (we will be within an exception during scheduling),
+	// and with interrupts enabled.
+	th->stack_state->sr = C0_STATUS() | C0_STATUS_EXL | C0_STATUS_IE;
+
+	// GP: this is needed by the compiler to point to the global memory pool,
+	// so let's initialize to the same value it has in this thread.
+	th->stack_state->gpr[28] = REG_GP();
+
+	// SP: top of stack
+	th->stack_state->gpr[29] = (int64_t)(int32_t)top_stack;
```

Comment:
Probably should be `= (intptr_t)top_stack;`

Reply:
`gpr[29]` is a `int64_t`, while `top_stack` is a 32-bit pointer. I need to convert it to `int64_t` in the same way MIPS assembly treats 32-bit values in 64-bit registers, which is by sign extending them. I don't want to rely on any property of integer promotion to casually match what the MIPS architecture does.